### PR TITLE
Remove unnecessary realpath calls from host probing logic 

### PR DESCRIPTION
**Pull request by mateoatr**
In an effort to improve startup time, this PR gets rid of three different file-accessing calls inside the host:

https://github.com/dotnet/runtime/blob/5265305d3121d2fce804948b9250b1b3a5f8a22d/src/native/corehost/hostpolicy/deps_resolver.cpp#L654

Which performs a file existence check on each of the assets inside a deps.json.

And:

https://github.com/dotnet/runtime/blob/8ad1d7d665c9867621193d4f8362cdb983602cae/src/native/corehost/hostpolicy/deps_resolver.cpp#L593

https://github.com/dotnet/runtime/blob/8ad1d7d665c9867621193d4f8362cdb983602cae/src/native/corehost/hostpolicy/deps_resolver.cpp#L48

Which purpose is to unroll the symbolic link of a given path and, in the later case, to later normalize the path.

### CODE REVIEW

File: src/native/corehost/hostpolicy/deps_resolver.cpp [hunk 1]
```diff
@@ -43,29 +43,25 @@ void add_unique_path(
     pal::string_t* non_serviced,
     const pal::string_t& svc_dir)
 {
-    // Resolve sym links.
-    pal::string_t real = path;
-    pal::realpath(&real);
-
-    if (existing->count(real))
+    if (existing->count(path))
```
Comment (vitek-karas) [inline src/native/corehost/hostpolicy/deps_resolver.cpp:46]: We should probably add a comment admitting that in presence of symlinks or non-normalized paths there might be effective duplicates in the output - and that it's OK.

File: src/installer/tests/Microsoft.NET.HostModel.Tests/Microsoft.NET.HostModel.AppHost.Tests/AppHostUsedWithSymbolicLinks.cs [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+                    .EnsureRestoredForRid(localizedFixture.CurrentRid)
+                    .PublishProject(runtime: localizedFixture.CurrentRid);
+
+                var publishFixture = new TestProjectFixture("StandaloneApp", RepoDirectories);
+                publishFixture
+                    .EnsureRestoredForRid(publishFixture.CurrentRid)
+                    .PublishProject(runtime: publishFixture.CurrentRid);
+
+                StandaloneAppFixture_Localized = localizedFixture;
+                StandaloneAppFixture_Published = publishFixture;
```
Comment (vitek-karas) [inline src/installer/tests/Microsoft.NET.HostModel.Tests/Microsoft.NET.HostModel.AppHost.Tests/AppHostUsedWithSymbolicLinks.cs:224]: Couple more tests we might want to add:
* Framework dependent app - run it from a symlink
* Framework dependent app where runtime install location is a symlink (basically `DOTNET_ROOT` points to a symlink)
* Running an app via `dotnet app.dll` where the `dotnet` is a symlink. (I think this is close to the snap scenario)

Comment (vitek-karas) [reply src/installer/tests/Microsoft.NET.HostModel.Tests/Microsoft.NET.HostModel.AppHost.Tests/AppHostUsedWithSymbolicLinks.cs:224]: Some more
* Make sure we have a test to validate additional probing paths (that they continue to work even after this change)

**Review by vitek-karas — APPROVED** [review]
Basically just some more interesting test cases, otherwise looks good.

File: src/native/corehost/deps_entry.h [hunk 3]
```diff
@@ -52,18 +52,18 @@ struct deps_entry_t
     bool is_rid_specific;
 
     // Given a "base" dir, yield the file path within this directory or single-file bundle.
-    bool to_dir_path(const pal::string_t& base, bool look_in_bundle, pal::string_t* str, bool& found_in_bundle) const;
+    bool to_dir_path(const pal::string_t& base, bool look_in_bundle, pal::string_t* str, bool& found_in_bundle, bool check_file_existence) const;
```
Comment (AaronRobinsonMSFT) [inline src/native/corehost/deps_entry.h:55]: Would strongly recommend keeping the last parameter as the out. In this case we now have "in in out out in". Also, we now have two booleans in these cases which means we are close to warranting an enumeration. At the very least though all of these functions should be updated to keep the outs as the last parameters.

**Review by brianrob — APPROVED** [review]

File: src/native/corehost/hostpolicy/deps_resolver.cpp [hunk 4]
```diff
@@ -304,6 +303,9 @@ bool deps_resolver_t::probe_deps_entry(const deps_entry_t& entry, const pal::str
             continue;
         }
         pal::string_t probe_dir = config.probe_dir;
+        uint32_t search_options = deps_entry_t::search_options::none;
+        if (m_has_additional_probing_paths || config.only_serviceable_assets)
+            search_options |= deps_entry_t::search_options::file_existence;
```
Comment (vitek-karas) [inline src/native/corehost/hostpolicy/deps_resolver.cpp:308]: I think we should change the logic here.
In `setup_probe_config` basically if we end up with anything but app and frameworks in the probes - we should enforce file existence checks. This applies to servicing, shared store, additional probing paths... and whatever else we add in the future.

**Review by vitek-karas — CHANGES_REQUESTED** [review]

**Review by vitek-karas — APPROVED** [review]

**Pull request merged**